### ISSUE TITLE
Feature Request: Performance Assessment Without Predefined NFRs

### ISSUE THREAD

**Issue by jdestevezcastillo-perfeng** [jdestevezcastillo-perfeng/Heimr.ai#18]
# Feature Request: Performance Assessment Without Predefined NFRs

## Summary

Enable Heimr to assess performance and make gate decisions **without requiring predefined NFRs** (Non-Functional Requirements). Support baseline mode, comparative mode, auto-discovery from production, industry presets, and user impact metrics.

## Problem Statement

**Current Heimr**: Requires explicit fail conditions to make gate decisions:
```bash
heimr agent results.json \
  --fail-condition "p99_latency > 500" \
  --fail-condition "error_rate > 1"
```

**Problem**: Most startups and small teams don't have formal NFRs. Common questions:
- "What's a good p99 latency for my API?"
- "How do I know if 300ms is acceptable?"
- "We just want to make sure we don't get slower — what threshold should we use?"

**This blocks adoption** for teams without performance engineering expertise.

## Proposed Solution

Add 5 new modes for performance assessment without predefined thresholds:

### 5. User Impact Mode

---

## 1. Baseline Mode

**Concept**: First test run establishes baseline. Future runs are compared against it.

### Usage:

**Establish baseline** (first run):
```bash
heimr agent results.json \
  --mode baseline \
  --save-baseline baseline.json
```

**Output**:
```
✅ Baseline established and saved to baseline.json

Performance metrics:
  - p99 latency: 320ms
  - p95 latency: 180ms
  - p50 latency: 95ms
  - Error rate: 0.3%
  - Throughput: 850 rps

Use --compare-to-baseline baseline.json in future runs.
```

**Compare against baseline** (subsequent runs):
```bash
heimr agent new-results.json \
  --compare-to-baseline baseline.json \
  --max-regression 20%
```

**Output**:
```
🔴 REJECT: Performance regression detected

Comparison to baseline:
  - p99 latency: 450ms (+40.6% from baseline 320ms) ⚠️ EXCEEDS 20% threshold
  - p95 latency: 210ms (+16.7% from baseline 180ms) ✅ Within threshold
  - Error rate: 0.3% (no change) ✅
  - Throughput: 820 rps (-3.5% from baseline 850 rps) ✅

Reasoning: p99 latency regression of 40.6% exceeds the maximum allowed
regression of 20%. This indicates a significant performance degradation.

Recommendation: Investigate recent code changes that may have introduced
latency. Check database query performance and external API calls.
```

### Implementation:

**baseline.json format**:
```json
{
  "version": "1.0",
  "created_at": "2026-03-08T10:30:00Z",
  "metrics": {
    "p99_latency_ms": 320,
    "p95_latency_ms": 180,
    "p50_latency_ms": 95,
    "error_rate_pct": 0.3,
    "throughput_rps": 850
  },
  "test_config": {
    "virtual_users": 100,
    "duration": "5m",
    "endpoints_tested": ["/api/users", "/api/products", "/api/checkout"]
  }
}
```

---

## 2. Comparative Mode

**Concept**: Compare current build against **previous build** or **main branch**.

### Usage:

```bash
# In CI/CD pipeline
heimr agent current-build-results.json \
  --compare-to previous-build-results.json \
  --max-regression 15%
```

**Alternative: Compare against main branch**:
```bash
# Test current PR
heimr agent pr-branch-results.json \
  --compare-to main-branch-results.json \
  --max-regression 10%
```

### GitHub Actions Integration:

```yaml
- name: Load test (main branch)
  run: k6 run load-test.js --out json=main-results.json

- name: Checkout PR branch
  uses: actions/checkout@v2

- name: Load test (PR branch)
  run: k6 run load-test.js --out json=pr-results.json

- name: Performance gate (comparative)
  uses: jdestevezcastillo-perfeng/heimr-ai@main
  with:
    results-file: pr-results.json
    compare-to: main-results.json
    max-regression: 15%
```

---

## 3. Auto-Discovery from Production

**Concept**: Query production observability systems to learn "normal" performance, then set test thresholds based on production reality.

### Usage:

```bash
heimr agent test-results.json \
  --discover-thresholds \
  --prometheus http://prometheus:9090 \
  --time-range 7d \
  --tolerance 20%
```

**What happens**:
1. Heimr queries Prometheus for the last 7 days of production metrics
2. Calculates production baseline: p99=280ms, error_rate=0.15%, throughput=1200rps
3. Sets test thresholds at +20% tolerance: p99<336ms, error_rate<0.18%, throughput>960rps
4. Evaluates test results against those thresholds

**Output**:
```
🔍 Auto-discovery: Querying production metrics (last 7 days)...

Production baseline:
  - p99 latency: 280ms (avg over 7d)
  - Error rate: 0.15%
  - Throughput: 1200 rps

📐 Test thresholds (production + 20% tolerance):
  - p99 latency: < 336ms
  - Error rate: < 0.18%
  - Throughput: > 960 rps

🔴 REJECT: Test environment performance worse than production

Test results vs production:
  - p99 latency: 450ms (⚠️ +60% worse than prod, exceeds 20% tolerance)
  - Error rate: 0.2% (⚠️ +33% worse than prod, exceeds 20% tolerance)
  - Throughput: 850 rps (⚠️ -29% worse than prod, exceeds 20% tolerance)

Reasoning: Test environment is significantly slower than production. This
suggests either inadequate test infrastructure or a real performance regression.

Recommendation: If test infra is known to be slower than prod, increase
--tolerance to 50%. Otherwise, investigate the performance degradation.
```

### Configuration:

```yaml
# heimr.yaml
auto_discovery:
  enabled: true
  prometheus:
    url: "http://prometheus:9090"
    time_range: "7d"
    metrics:
      - name: "http_request_duration_seconds"
        quantile: 0.99
        label_filters: {service: "api"}
      - name: "http_requests_total"
        label_filters: {service: "api", status: "5xx"}
  tolerance: 20%  # Test can be 20% worse than prod
```

---

## 4. Industry Preset Profiles

**Concept**: Use research-backed, industry-standard thresholds for common application types.

### Usage:

```bash
heimr agent results.json --preset web-api
heimr agent results.json --preset ecommerce
heimr agent results.json --preset realtime
heimr agent results.json --preset microservices
```

### Preset Definitions:

**web-api** (General REST APIs):
```yaml
thresholds:
  p99_latency: 500ms
  p95_latency: 300ms
  error_rate: 1%
  availability: 99.9%
source: "Industry standard for B2B APIs"
```

**ecommerce** (E-commerce websites/APIs):
```yaml
thresholds:
  product_page_load: 2s
  checkout_flow: 3s
  search: 200ms
  error_rate: 0.5%
source: "Google Web Vitals + Amazon research (100ms = 1% revenue)"
```

**realtime** (Chat, gaming, collaborative tools):
```yaml
thresholds:
  p99_latency: 100ms
  p95_latency: 50ms
  error_rate: 0.1%
  websocket_message_delay: 50ms
source: "Human perception research (<100ms = instant)"
```

**microservices** (Internal service-to-service):
```yaml
thresholds:
  p99_latency: 200ms
  p95_latency: 100ms
  error_rate: 0.5%
  circuit_breaker_threshold: 5%
source: "Google SRE Book"
```

**mobile-api** (Mobile app backends):
```yaml
thresholds:
  p99_latency: 1000ms  # Mobile networks are slower
  error_rate: 2%       # Handle flaky connections
  offline_capable: true
source: "Mobile-first performance standards"
```

### Research-Backed Justification:

Each preset includes citations:
- **Google**: 53% of mobile users abandon sites >3s ([Source](https://www.thinkwithgoogle.com/marketing-strategies/app-and-mobile/mobile-page-speed-new-industry-benchmarks/))
- **Amazon**: 100ms latency = 1% sales loss ([Source](https://www.gigaspaces.com/blog/amazon-found-every-100ms-of-latency-cost-them-1-in-sales))
- **Akamai**: 2s delay = 103% bounce rate increase
- **Nielsen**: <100ms perceived as "instant", 1s = flow interrupted

---

## 5. User Impact Mode

**Concept**: Translate technical metrics into **business impact** (abandonment rate, conversion loss, revenue impact).

### Usage:

```bash
heimr agent results.json \
  --user-impact-mode \
  --acceptable-abandonment-rate 5% \
  --acceptable-conversion-loss 2%
```

### How it works:

Uses research-backed formulas to estimate business impact:

**Abandonment rate formula** (Google research):
```
abandonment_rate = 0.53 * (load_time_seconds / 3.0) ^ 2
```

**Conversion loss formula** (Amazon research):
```
conversion_loss_pct = (latency_ms - baseline_latency) / 100
```

**Revenue impact** (if --revenue-per-user provided):
```
revenue_loss = conversion_loss_pct * users_per_day * revenue_per_user
```

### Output:

```
📊 User Impact Analysis

Technical metrics:
  - p99 latency: 4.2s
  - Error rate: 0.8%

Estimated business impact:
  - Abandonment rate: 9.8% ⚠️ EXCEEDS 5% threshold
  - Conversion loss: ~3.2% ⚠️ EXCEEDS 2% threshold
  - Estimated daily revenue loss: $4,500 (based on 10k users/day @ $14 avg order)

🔴 REJECT: Performance likely to cause significant user abandonment

Reasoning: With a p99 load time of 4.2 seconds, approximately 10% of users
will abandon before page load completes. This exceeds your acceptable
abandonment threshold of 5%.

Recommendation: Optimize to reduce p99 below 2 seconds to keep abandonment
rate below 5%.
```

---

## Minimum Requirements

To use these modes, users need:

### Baseline Mode:
- At least one successful load test run
- Writable filesystem to save baseline.json

### Comparative Mode:
- Results from two test runs (current + previous)

### Auto-Discovery:
- Access to production Prometheus/observability
- At least 24 hours of production metrics (7d recommended)

### Industry Presets:
- Nothing! Just `--preset <name>`

### User Impact Mode:
- Load test results with latency data
- Optional: `--revenue-per-user` for revenue impact calculation

---

## Acceptance Criteria

- [ ] Baseline mode: `heimr agent --mode baseline --save-baseline <path>` works
- [ ] Baseline comparison: `heimr agent --compare-to-baseline <path>` works
- [ ] Regression threshold: `--max-regression <pct>` configurable (default 20%)
- [ ] Comparative mode: `heimr agent --compare-to <results.json>` works
- [ ] Auto-discovery: `heimr agent --discover-thresholds --prometheus <url>` works
- [ ] Auto-discovery respects `--time-range` (1d, 7d, 30d)
- [ ] Auto-discovery respects `--tolerance` (default 20%)
- [ ] Industry presets: `--preset web-api` works
- [ ] At least 5 presets: web-api, ecommerce, realtime, microservices, mobile-api
- [ ] Each preset includes research citations in output
- [ ] User impact mode: `--user-impact-mode` works
- [ ] Abandonment rate calculation accurate
- [ ] Conversion loss calculation accurate
- [ ] Revenue impact calculation (if `--revenue-per-user` provided)
- [ ] All modes produce full audit trail (like current agent)
- [ ] All modes work with existing `--gate-policy` (strict, moderate, lenient)
- [ ] Documentation updated with examples for each mode

---

### Baseline Storage:

Store baselines in a `.heimr/` directory by default:
```
.heimr/
├── baselines/
│   ├── main-branch.json
│   ├── staging.json
│   └── production-discovery.json
└── config.yaml
```

Git-ignore `.heimr/baselines/` by default (user can commit if desired).

### Auto-Discovery Caching:

Cache production metrics for 1 hour to avoid hammering Prometheus:
```
.heimr/cache/
└── prod-metrics-2026-03-08T10-00.json
```

### Preset Extensibility:

Allow users to define custom presets in `heimr.yaml`:
```yaml
presets:
  my-custom-api:
    p99_latency: 400ms
    error_rate: 0.5%
    source: "Our internal SLA"
```

---

## Related Work

- [Deployment Gating](https://heimr.ai/docs.html?doc=02-deployment-gating) — Current agent mode
- [Configuration Guide](https://heimr.ai/docs.html?doc=08-configuration) — Threshold configuration
- Google Web Vitals — User-centric performance metrics
- Google SRE Book — SLO/SLI/error budget concepts

---

## Prior Art

- **Lighthouse CI**: Uses baseline comparison for web performance
- **k6 Cloud**: Stores baseline and compares test runs
- **New Relic**: Auto-discovery of baseline from APM data
- **Google Web Vitals**: User impact thresholds (LCP, FID, CLS)

---

## Community Interest

This issue was inspired by a discussion with a DevOps engineer who asked:

> "We don't have NFRs defined. How can we assess if our performance is good?"

**This is a common pain point for:**
- Startups without formal performance requirements
- Teams transitioning from "no testing" to "some testing"
- DevOps engineers who know infra but not performance engineering
- Small teams without dedicated QA/performance engineers

---

## Questions for Maintainers

1. Should baseline mode be the **default** if no fail-conditions are provided?
2. Should auto-discovery require explicit opt-in or be suggested automatically?
3. Should presets be hardcoded or loaded from a YAML file (extensibility)?
4. Should we integrate with Google Web Vitals directly for frontend tests?
5. Should revenue impact calculations be opt-in or always shown?

---

## Additional Context

**This makes Heimr accessible to non-experts.** Currently, you need to know what "p99 < 500ms" means. With these modes, you can:
- Start with a baseline (no expertise needed)
- Use industry presets (trust research-backed standards)
- Auto-discover from production (align test with reality)
- Understand business impact (translate to revenue/users)

**This dramatically lowers the barrier to entry** for performance testing.

---

**Would this feature be valuable to you?** 👍 this issue or comment with your use case!

For commercial development or priority implementation:
- Contact: jd.estevezcastillo@gmail.com
